mypy/linting: null=true implies optional

neiljp opened the pull request:
After reading many annotations in `models.py` (and exmaining one test case, whose exclusions are modified in one of these commits), I noticed that typically 'null=true' implies an Optional.

Assuming this is a reasonable assumption, this PR makes this connection explicit, adding a linter check and modifying annotations accordingly. One strict-optional exclusion is modified, since the test now passes but another file does not, and is left for later to resolve.

One associated issue is whether a linter rule should be added for the inverse, namely presence of Optional should also imply null=true.

timabbott commented:
I think a lint rule for the opposite would also be correct.

timabbott commented on tools/linter_lib/custom_check.py:
```diff
@@ -484,6 +484,11 @@ def build_custom_checkers(by_lang):
+        {'pattern': 'null=True',
+         'exclude_pattern': '# type: Optional',
+         'include_only': {"zerver/models.py"},
+         'description': 'Model variable with null=true not annotated as Optional.',
```
Can you add a `good_lines` and `bad_lines` to test this new rule?

We're sorta working on a goal of having all of our linter rules have tests.

timabbott commented on tools/linter_lib/custom_check.py:
```diff
@@ -484,6 +484,11 @@ def build_custom_checkers(by_lang):
          ]),
          'description': 'Comment-style function type annotation. Use Python3 style annotations instead.',
          },
+        {'pattern': 'null=True',
```
I might write this as `= models[.].*null=True` to make it a bit less spammy

neiljp replied:
I've also rewritten the original check to avoid the `exclude_pattern`, which appears to still work, but I couldn't get that working for the combination in the second form.

neiljp commented:
* Added additional linter rule, which takes NullBooleanField into account & one change to `models.py`;
* Amended pattern for first linter rule to be more specific;
* Added `good_lines` and `bad_lines` tests for both linter rules.

Passed `test-all` on droplet.

neiljp commented:
@timabbott Should be good now, FYI.

timabbott commented:
Looks great, merged, thanks @neiljp!
